internal/provider/user_role_grant_resource.go

```diff
@@ -0,0 +1,272 @@
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type userRoleGrantResource struct {
+	provider *provider
+}
+
+func (r *userRoleGrantResource) Schema(
+	_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse,
+) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "A role grant for a user. This resource is recommended to be used when a user's roles are managed across multiple terraform projects or in conjunction with console UI granted roles. For authoritative management over a user's roles, use the [user_role_grants](user_role_grants) resource.\n\n As with all terraform resources, care must be taken to limit management of the same resource to a single project.",
+		Attributes: map[string]schema.Attribute{
+			"user_id": schema.StringAttribute{
+				Required:    true,
+				Description: "ID of the user to grant these roles to.",
+			},
+			"role": schema.SingleNestedAttribute{
+				Required: true,
+				Attributes: map[string]schema.Attribute{
+					"role_name": schema.StringAttribute{
+						Required: true,
+						MarkdownDescription: "Name of the role to grant. Allowed values are:" +
+							formatEnumMarkdownList(client.AllowedOrganizationUserRoleTypeEnumValues),
+					},
+					"resource_type": schema.StringAttribute{
+						Required: true,
+						MarkdownDescription: "Type of resource. Allowed values are:" +
+							formatEnumMarkdownList(client.AllowedResourceTypeTypeEnumValues),
+					},
+					"resource_id": schema.StringAttribute{
+						Optional:    true,
+						Description: "ID of the resource. Omit if resource_type is 'ORGANIZATION'.",
+					},
+				},
+			},
+		},
+	}
+}
+
+func (r *userRoleGrantResource) Metadata(
+	_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse,
+) {
+	resp.TypeName = req.ProviderTypeName + "_user_role_grant"
+}
+
+func (r *userRoleGrantResource) Configure(
+	_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse,
+) {
+	if req.ProviderData == nil {
+		return
+	}
+	var ok bool
+	if r.provider, ok = req.ProviderData.(*provider); !ok {
+		resp.Diagnostics.AddError("Internal provider error",
+			fmt.Sprintf("Error in Configure: expected %T but got %T", provider{}, req.ProviderData))
+	}
+}
+
+func (r *userRoleGrantResource) Create(
+	ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse,
+) {
+	if r.provider == nil || !r.provider.configured {
+		addConfigureProviderErr(&resp.Diagnostics)
+		return
+	}
+
+	var plan UserRoleGrant
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	_, err := fromRoleToBuiltInRole(plan.Role)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Invalid role format",
+			fmt.Sprintf("Could not parse role: %s", formatAPIErrorMessage(err)),
+		)
+		return
+	}
+	_, _, err = r.provider.service.AddUserToRole(
```

Comment:
Do we need to do any logic here to check if this assignment is necessary? Or should we just always do it because it's idempotent?

Reply:
In general I'm not concerned with idempotency as terraform sprinkles read in various places to determine if something is necessary. Drift can obviously happen and whether it happens during or close to a terraform run is just luck.  It would be detected during the next run and updated.   For this particular case however I can't recall if I've tested what happens when a role grant already exists.  It *should* fail in this case but if our API is idempotent, it may think that it has been able to "adopt" this resource without using import.   I'll do some testing around this.  Maybe I can add an integration test.

Reply:
So I think this is actually a problem.  If terraform tries to create a resource which already exists and doesn't fail, it means two projects that are accidentally using the same resource will both think they own it.

Reply:
It could also lead to issues in the same project where a project thinks it has 2 managed resources but only one on the backend.

Reply:
I pushed an update to check for the resource before attempting to "Create" it.  I've also updated the tests to exercise a failure.